docs/docs/vm-fundamentals/builtins.md

```diff
@@ -4,4 +4,229 @@ sidebar_position: 2
+
+Imagine the VM memory as a filing cabinet, builtins are like add-on modules that create special folders within the cabinet and each builtin gets its own folder and dictates the rules for what can be stored there. For instance, a "range-check" builtin might create a folder where all values must be between 0 and a very large number.
+
+**Communication between the CPU and builtins happens entirely through this memory.** 
+
+The CPU doesn't directly call builtins, instead, it treats the builtins memory folder like a special device to use a builtin. A function gets a pointer to the builtins folder as an argument, the function then reads or writes data to specific cells within the folder, following the rules set by the builtin. This method is similar to how external devices interact with a computer, they don't talk directly to the CPU but use the memory space for communication.
+
+**Adding builtins doesn't require modifying the CPU itself.** 
+
+Both the CPU and builtins share the same memory space, making them efficient tools. Additionally, each function using a builtin receives a pointer and is responsible for returning a pointer to the next unused cell within the builtin's memory area. This ensures proper management of space allocated for each builtin instance.
+
+### General types of builtins in Cairo VM
+
+- **Mathematical builtins.**
+    
+    These perform various mathematical operations, potentially faster than using regular Cairo instructions. Examples include addition, subtraction, multiplication, bitwise operations (AND, OR, XOR, etc.), and some even handle more complex operations like modular exponentiation.
+    
+- **Cryptographic builtins.**
+    
+    These handle cryptographic functions used in zk-STARK proofs. Examples include functions for hashing (like Keccak) and various elliptic curve cryptography operations. Some examples are:
+    
+    **→ Keccak hashing:**
+    
+    This builtin performs the Keccak hash function, a one-way function that creates a unique and fixed-size output from any input data. It's used to compress data and ensure its integrity in zk-STARK proofs.
+    
+    **→ Scalar multiplication:**
+    
+    This builtin performs point multiplication on elliptic curves, a fundamental operation in elliptic curve cryptography (ECC). ECC is used for secure communication and signature verification in zk-STARKs.
+    
+- **Memory management builtins.**
+    
+    These help manage memory allocation and deallocation within a Cairo program. This can be crucial for optimizing memory usage. Some examples are:
+    
+    **→ Memory allocation:**
+    
+    This builtin allocates a specific amount of memory within the Cairo VM for a program to use. This ensures the program doesn't try to access non-existent memory locations.
+    
+    **→ Memory deallocation:**
+    
+    This builtin frees up memory that is no longer needed by a program. This prevents memory leaks and optimizes memory usage for subsequent computations.
```

Comment:
Memory in Cairo is write-once, so there is never actually any deallocation. All memory written stays like that forever.